### PR TITLE
Structural/adding spectral decomposition method to CL->utilities

### PR DESCRIPTION
**Pull request by AlejandroCornejo**

### CODE REVIEW

**Comment by AlejandroCornejo** [comment]
Anyone knows why the travis has failed?? @loumalouomega

**Comment by loumalouomega** [comment]
Travis transforms some warnings into errors

**Review by loumalouomega — APPROVED** [review]

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp [hunk 1]
```diff
@@ -523,6 +523,205 @@ void ConstitutiveLawUtilities<TVoigtSize>::CalculatePrincipalStressesWithCardano
...
+    Vector cross_p_ij_vector;
+
+    for (IndexType index = 0; index < Dimension; ++index) {
+        i = indexes_ij(index, 0);
+        j = indexes_ij(index, 1);
+
+        if (eigen_values_matrix(i, i) > 0)
+            h_i = 1.0;
+        if (eigen_values_matrix(j, j) > 0)
+            h_j = 1.0;
```
Comment (tteschemacher) [inline applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:585]: These h_i, h_j should be set to zero after every iteration, right?

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp [hunk 2]
```diff
@@ -523,6 +523,205 @@ void ConstitutiveLawUtilities<TVoigtSize>::CalculatePrincipalStressesWithCardano
...
+    indexes_ij(1, 1) = 2;
+    indexes_ij(2, 0) = 0;
+    indexes_ij(2, 1) = 2;
+
+    IndexType i, j;
+    double h_i = 0.0, h_j = 0.0;
+    Matrix cross_p_ij_tensor;
+    Vector cross_p_ij_vector;
+
+    for (IndexType index = 0; index < Dimension; ++index) {
```
Comment (tteschemacher) [inline applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:578]: In the publication from cevera this is not written very clean in this formulary. But I think this loop is: sum (index from 0 to dimension) sum (from index to dimension)

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp (hunk 1 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:585]: Sorry I already merged, if any change is required a new PR should be done

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:578]: Same here

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp (hunk 1 above)
Comment (tteschemacher) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:585]: Sure I saw it. Let's see first if we will need any changes.

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:585]: I'm correcting the erorrs in another branch do not worry

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:585]: thank you @tteschemacher  you're totally right

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:578]: mm okay maybe I haven't understood correctly this sum.. changing it!

**Pull request merged**